### PR TITLE
[TSE] Implement the Exchange Data Service

### LINKED ISSUE #7: [TSE] Implement the Exchange Data Service
### Description
This service will be responsible for gathering data from the different supported exchanges[CCTX]. It will retrieve market data either through REST APIs or WebSockets from these exchanges.
[Example](https://github.com/Hu-Fi/Mr.Market/blob/main/server/src/modules/marketdata/marketdata.service.ts)

### To Do
- Implement `getTickers` method: Fetches tickers for specified symbols from a given exchange
- Implement `getOHLCVData` method: Fetches OHLCV data for a specified symbol from a given exchange
- Implement `getSupportedPairs` method: Fetches supported trading pairs
- Implement `watchOrderBook` method: Subscribes to order book updates for a specified symbol from a given exchange
- Implement `watchOHLCV` method: Subscribes to OHLCV data updates for a specified symbol from a given exchange
- Implement `watchTicker` method: Subscribes to ticker updates for a specified symbol from a given exchange
- Implement `watchTickers` method: Subscribes to ticker updates for multiple symbols from a given exchange
- Implement `isSubscribed` method: Checks if a subscription is active for the given parameters
- Implement `unsubscribeOrderBook` method: Unsubscribes from order book updates for a specified symbol
- Implement `getTickerPrice` method: Fetches the latest ticker price for a specified symbol from a given exchange
- Implement `getMultipleTickerPrices` method: Fetches ticker prices for multiple symbols across multiple exchanges
- Implement `getSupportedSymbols` method: Fetches supported symbols from a given exchange
- Implement `unsubscribeData` method: Unsubscribes from data updates for the given parameters

### Deliverables
- [ ] All tasks completed
- [ ] Unit tests
- [ ] Documentation

### PR DESCRIPTION
**Pull request by MWBlocky**
## Description

This Pull Request introduces the implementation of various methods to gather and manage market data from different supported exchanges using CCXT. The service retrieves market data through REST APIs or WebSockets from these exchanges. The following methods are implemented to handle fetching and subscribing to market data:

## Summary of changes

- **Implemented `getTickers` method**: Fetches tickers for specified symbols from a given exchange.
- **Implemented `getOHLCVData` method**: Fetches OHLCV data for a specified symbol from a given exchange.
- **Implemented `getSupportedPairs` method**: Fetches supported trading pairs from exchanges.
- **Implemented `watchOrderBook` method**: Subscribes to order book updates for a specified symbol from a given exchange.
- **Implemented `watchOHLCV` method**: Subscribes to OHLCV data updates for a specified symbol from a given exchange.
- **Implemented `watchTicker` method**: Subscribes to ticker updates for a specified symbol from a given exchange.
- **Implemented `watchTickers` method**: Subscribes to ticker updates for multiple symbols from a given exchange.
- **Implemented `isSubscribed` method**: Checks if a subscription is active for the given parameters.
- **Implemented `unsubscribeOrderBook` method**: Unsubscribes from order book updates for a specified symbol.
- **Implemented `getTickerPrice` method**: Fetches the latest ticker price for a specified symbol from a given exchange.
- **Implemented `getMultipleTickerPrices` method**: Fetches ticker prices for multiple symbols across multiple exchanges.
- **Implemented `getSupportedSymbols` method**: Fetches supported symbols from a given exchange.
- **Implemented `unsubscribeData` method**: Unsubscribes from data updates for the given parameters.
- **Added unit tests**: Ensured that all methods function correctly and handle edge cases as expected.
- **Documentation**: Updated documentation to reflect the new methods and their usage.

## Related issues

Fixes #7

### CODE REVIEW

File: packages/tse-backend/src/modules/exchange-data/exchange-data.mapper.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  GetTickersDto,
+} from './model/exchange-data.model';
+
+@Injectable()
+export class ExchangeDataProfile extends AutomapperProfile {
+  constructor(@InjectMapper() mapper: Mapper) {
+    super(mapper);
+  }
+
+  override get profile() {
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-data/exchange-data.mapper.ts:23]: What's that?, why do we need to ovverride it?

File: packages/tse-backend/src/common/utils/subscriptionKey.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    case MarketDataType.TICKER:
+      decodedKey.symbol = parts[2];
+      break;
+    case MarketDataType.OHLCV:
+      decodedKey.symbol = parts[2];
+      decodedKey.timeFrame = parts[3];
+      break;
+    case MarketDataType.TICKERS:
+      decodedKey.symbols = parts[2].split(',');
+      break;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/common/utils/subscriptionKey.ts:46]: Why do we need this switch?

Comment (piotrswierzy) [reply packages/tse-backend/src/common/utils/subscriptionKey.ts:46]: Can't we implement a strategy pattern insted of it?

File: packages/tse-backend/src/common/utils/subscriptionKey.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+): string => {
+  let key = '';
+  if (type === MarketDataType.ORDERBOOK || type === MarketDataType.TICKER) {
+    key = `${type}:${exchange}:${symbol}`;
+  } else if (type === MarketDataType.OHLCV) {
+    key = `${type}:${exchange}:${symbol}:${timeFrame}`;
+  } else if (type === MarketDataType.TICKERS) {
+    key = `${type}:${exchange}:${symbols}`;
+  }
+  return key;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/common/utils/subscriptionKey.ts:26]: What's that, and why do we need this?

File: packages/tse-backend/src/integrations/ccxt.gateway.ts [hunk 4]
```diff
@@ -1,37 +1,43 @@
...
+      }
+      const exchange = new exchangeClass({
+        apiKey,
+        secret,
+      });
+      await exchange.loadMarkets();
+      return exchange;
+    } catch (error) {
+      this.logger.warn(`Failed to initialize ${name}: ${error.message}`);
+      return null;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/integrations/ccxt.gateway.ts:40]: Shouldn't we throw an Error instead of returning null?

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+  async getTickers(command: GetTickersCommand) {
+    const { exchange, symbols } = command;
+    const exchangeInstance = this.exchangeRegistryService.getExchange(exchange);
+    if (!exchangeInstance || !exchangeInstance.has.fetchTickers) {
+      throw new Error(
+        'Exchange does not support fetchTickers or is not configured.',
+      );
+    }
+
+    const symbolsArray = symbols.split(',');
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts:38]: Why symbols is a string instead of an list?

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+      if (exchangeInstance && exchangeInstance.has.fetchTickers) {
+        try {
+          const tickers = await exchangeInstance.fetchTickers();
+          pairs.push(...Object.keys(tickers));
+        } catch (error) {
+          this.logger.error(
+            `Error fetching tickers from ${exchange}: ${error}`,
+          );
+        }
+      }
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts:107]: Extract this to a separate private method

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    this.logger.log(
+      `Fetching supported pairs from ${exchangeInstances.length} exchanges`,
+    );
+
+    for (const exchange of exchangeInstances) {
+      const exchangeInstance =
+        this.exchangeRegistryService.getExchange(exchange);
+      if (exchangeInstance && exchangeInstance.has.fetchTickers) {
+        try {
+          const tickers = await exchangeInstance.fetchTickers();
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts:100]: This should probably be cached if it's an API call, if so add an //TODO comment

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+  async getSupportedSymbols(command: GetSupportedSymbolsCommand) {
+    const { exchange } = command;
+    const exchangeInstance = this.exchangeRegistryService.getExchange(exchange);
+    if (!exchangeInstance) {
+      throw new Error(`Exchange ${exchange} is not configured.`);
+    }
+    this.logger.log(`Fetching supported symbols from ${exchangeInstance.name}`);
+
+    try {
+      await exchangeInstance.loadMarkets();
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts:181]: This can also probably be cached if it's an API call, if so add an //TODO comment as well

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+          symbol || symbols,
+          ...Object.values(extraParams),
+        );
+        onData(data);
+      }
+    } catch (error) {
+      this.logger.error(
+        `Error watching ${type} for ${symbol || symbols} on ${exchange}: ${error.message}`,
+      );
+      await new Promise((resolve) => setTimeout(resolve, 1000)); // Reconnect after a delay
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts:226]: What's the purpose of that?

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+
+  async watchOrderBook(
+    exchange: string,
+    symbol: string,
+    onData: (data: any) => void,
+    limit = 14,
+  ): Promise<void> {
+    if (exchange === 'bitfinex') {
+      limit = 25;
+    }
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts:238]: Why do we need this?

File: packages/tse-backend/src/modules/exchange-data/model/exchange-data.model.ts [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  @IsString()
+  exchange: string;
+}
+
+export class GetTickersCommand {
+  @AutoMap()
+  exchange: string;
+
+  @AutoMap()
+  symbols: string;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-data/model/exchange-data.model.ts:86]: Why is this a string and not a list?

File: packages/tse-backend/src/modules/exchange-data/model/exchange-data.model.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+export class GetSupportedSymbolsDto {
+  @AutoMap()
+  @ApiProperty({ example: 'binance' })
+  @IsString()
+  exchange: string;
+}
+
+export class GetTickersCommand {
+  @AutoMap()
+  exchange: string;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-data/model/exchange-data.model.ts:83]: Why isn't this an Enum?

**Review by piotrswierzy — CHANGES_REQUESTED** [review]

File: packages/tse-backend/src/modules/exchange-data/exchange-data.mapper.ts (hunk 1 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.mapper.ts:23]: https://automapperts.netlify.app/docs/nestjs#automapperprofile

File: packages/tse-backend/src/common/utils/subscriptionKey.ts (hunk 2 above)
Comment (MWBlocky) [reply packages/tse-backend/src/common/utils/subscriptionKey.ts]: fixed: ca556650

File: packages/tse-backend/src/common/utils/subscriptionKey.ts (hunk 3 above)
Comment (MWBlocky) [reply packages/tse-backend/src/common/utils/subscriptionKey.ts]: This utility automatically creates and manages unique keys based on subscriber-defined data preferences, simplifying subscription management and data routing in websocket

File: packages/tse-backend/src/integrations/ccxt.gateway.ts (hunk 4 above)
Comment (MWBlocky) [reply packages/tse-backend/src/integrations/ccxt.gateway.ts]: Fixed: 8d59f40e

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts (hunk 5 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts]: Fixed: ce69c8d6

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts (hunk 6 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts]: Fixed: bb0c4fb3

File: packages/tse-backend/src/modules/exchange-data/model/exchange-data.model.ts (hunk 11 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/model/exchange-data.model.ts]: Fixed: https://github.com/Hu-Fi/Mr.Market-V2/commit/ce69c8d69353d82634554871e41ebae3ebaef0d9

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts (hunk 8 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts:181]: I agree it should be cached: 73d8b9b4

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts (hunk 7 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts]: I agree it should be cached: https://github.com/Hu-Fi/Mr.Market-V2/commit/73d8b9b44ac019d85a18c37a064e3b5f804ea7b6

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts (hunk 10 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts:238]: Bitfinex requires the watchOrderBook limit parameter to be set to either undefined, 25, or 100, as part of their API constraints

File: packages/tse-backend/src/modules/exchange-data/model/exchange-data.model.ts (hunk 12 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/model/exchange-data.model.ts:92]: In our API, any exchange name can be passed, and if it's not configured (keys missing in env), the API returns a message; while ccxt lib could be used, we avoid this to keep modules separate and imports clean

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts (hunk 9 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts:226]: In cases where the ccxt API is temporarily unavailable, we don't want to exit the loop; instead, we want to keep receiving data. This delay helps us avoid sending too many requests. However, since they don't use this delay in their example (https://github.com/ccxt/ccxt/blob/master/examples/ccxt.pro/js/gateio-watch-order-book.js#L31), and it may not be necessary now, I'll remove this line.

Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts]: fix: 1fb1dc4e

File: packages/tse-backend/src/common/utils/subscriptionKey.ts (hunk 3 above)
Comment (piotrswierzy) [reply packages/tse-backend/src/common/utils/subscriptionKey.ts]: add a code comment explaining it

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts (hunk 9 above)
Comment (piotrswierzy) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts]: Did you test it, if it's not necessary?

File: packages/tse-backend/src/common/utils/subscriptionKey.ts (hunk 3 above)
Comment (MWBlocky) [reply packages/tse-backend/src/common/utils/subscriptionKey.ts]: Added: 7fdad0b8

File: packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts (hunk 9 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-data/exchange-data.service.ts]: It will be hard to test this. I checked the forums and didn't find anyone who had a problem with the ban. As for exchange restrictions, CCXT has a built-in rate limiter that automatically manages the request speed.

**Review by piotrswierzy — APPROVED** [review]
